### PR TITLE
support brushnet (BrushNet: A Plug-and-Play Image Inpainting Model with Decomposed Dual-Branch Diffusion)

### PR DESCRIPTION
**Pull request by juxuan27**
# What does this PR do?

This PR is to support [BrushNet](https://github.com/TencentARC/BrushNet) (BrushNet: A Plug-and-Play Image Inpainting Model with Decomposed Dual-Branch Diffusion) in diffusers. It is a plug-and-play inpainting method that have been accepted to ECCV2024.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/brushnet/train_brushnet.py [hunk 1]
```diff
@@ -0,0 +1,1403 @@
...
+from packaging import version
+from PIL import Image, ImageDraw
+from torchvision import transforms
+from tqdm.auto import tqdm
+from transformers import AutoTokenizer, PretrainedConfig
+
+import diffusers
+from diffusers import (
+    AutoencoderKL,
+    BrushNetModel,
```
Comment (cs-mshah) [inline examples/brushnet/train_brushnet.py:38]: We need to add the model classes as well.

**Review by cs-mshah — CHANGES_REQUESTED** [review]
We need to add the `BrushNetModel` and `StableDiffusionBrushNetPipeline` to diffusers as well.

**Comment by cs-mshah** [comment]
Would it be a better idea  to decouple the code and keep it only to the `examples` folder as done by https://github.com/attashe/BrushNetSimple

**Comment by juxuan27** [comment]
I am working on this! Sorry for forgetting to add the `BrushNetModel` and `StableDiffusionBrushNetPipeline`. I will pin you once I finished. Thanks for your review.

**Comment by yiyixuxu** [comment]
I think this should either go into the community folder (https://github.com/huggingface/diffusers/tree/main/examples/community) as a community pipeline, or research_projects folder (https://github.com/huggingface/diffusers/tree/main/examples/research_projects) , no?

**Pull request closed without merging**